week-07-rails-continued/day-01/drill/readme.md

```diff
@@ -1 +1,110 @@
+
+### Binary Search Trees
+
+Binary search trees add on an extra restriction to binary trees: each node's left child subtree (if it has one) will  contain only nodes whose values are lower than the original node's value.  In each node's right child subtree (if it has one), all nodes must have a greater value than the original node itself.
+
+
+<img src="images/bst.jpg" width="300px">
+
+
+### Balanced Binary Trees
+
+Balanced binary trees are another basic variant of binary trees. A "balanced" tree has a height about as low as it can possibly be while still holding all its nodes.  For binary trees, that means the height is O(log<sub>2</sub>(n)), where n is the number of nodes in the tree.  There are different definitions of exactly how to balance a tree, but you can tell a tree is balanced if all of the leaves are either at the very bottom level of the tree or just one level higher.
+
+### Balanced Binary Search Trees
+
+Balanced binary search trees combine the balanced structure requirement with the node value requirement of binary search trees.  If an interview question asks about a tree, try to clarify whether the tree is balanced and whether it is a binary search tree.
+
+
+### Tries
+
+Tries, also called prefix trees, aren't usually binary.  They allow each node to have as many children as needed. The special thing about tries is how they store data. The data builds up over the path from the root to each node.  Here's an example:
+
+<img src="images/trie.jpg" width="300px">
+
+## Challenges
+
+PLEASE DO NOT CODE UNLESS A CHALLENGE SPECIFICALLY INSTRUCTS YOU TO
+
+Assume for the following challenges that you have a `binary_tree` data structure allowing you to:
+
+* access the root node with `.root`.
+* given any node, find the left child of that node with `.left_child`
+* given any node, find the right child of that node with `.right_child`
+* given any node, find the parent of that node with `.parent`
+
+Also assume a `trie` data structure that allows you to:
+
+* find the root with `.root`
+* given any node, get a list of the node's children with `.children`
+* given any node, find the node's parent with `.parent`
+
+
+### Challenges - Binary Search Trees
+
+1. Create a binary search tree from the following array: [0,1,2,3,4,5,6].
+
+2. Describe an algorithm to check if a particular number value is inside a binary search tree.  *Hint: start by checking if it's the value of the root.*
+
+3. In a binary search tree, how can you find the minimum element? The maximum?
+
+4. Assume for the following challenges that you have a binary search tree data structure. The structure will be recursive, meaning each node is actually represented as a full subtree, an instance of `BinarySearchTree`.
```

Comment:
@ben-manning The solution for this challenge is about hex colors - is that what this challenge should be? They don't quite match up.